spec/p2p/p2p.qnt

```diff
@@ -0,0 +1,577 @@
+
+/*
+ * Useful actions to simplify spec
+ */ 
+
+    action allUnchanged = all {
+        config' = config,
+        Environment' = Environment,
+        OperatorAddressBook' = OperatorAddressBook,
+        requestingNodes' = requestingNodes,
+    }    
+
+    // Framerule
+    action environmentUnchanged = all {
+        Environment' = Environment,
+        OperatorAddressBook' = OperatorAddressBook,
+        requestingNodes' = requestingNodes,
+    }    
+
+    
+
+/*
+ * Local actions
+ */
+
+   // node dials a. This starts lower lever computations. 
+    action actionDialPeer (node: Address, a:Address):bool = all {
+        config' = config.set(node, config.get(node).dialPeer(a)),
+        environmentUnchanged,
+    }
+
+    action actionTransportAccept (node: Address, a:Address):bool = all {
+        config' = config.set(node, config.get(node).transportAccept(a)),
+        environmentUnchanged,
+    }
+
+    action actionOnLostConnection(node: Address, c: Connection): bool = all {
+        config' = config.set(node, config.get(node).onLostConnection(c)),
+        environmentUnchanged,
+    }
+
+/* 
+ * Environment actions: From the lower level (modeling the outcome of TCP connection
+ * initialization, establishment, etc.)
+ */
+
+    // auxiliary functions
+    pure def removeConnectionOutgoing(s: LocalState, a: Address): LocalState = {
+        val y : Set[Connection] = s.connections.exclude(Set({address: a, state: "outgoing"}))
+        s.with("connections", y)
+    }
+
+    pure def putPeer(s: LocalState, a: Address, direction: str): LocalState = {
+        val y =  s.peerState.put(a, direction)
+        s.with("peerState", y)
+    }
+
+    pure def loseConnection(s: LocalState, a: Address): LocalState = {
+        val y : Set[Connection] = 
+            s.connections.exclude(Set({address: a, state: "outgoing"}, {address: a, state: "incoming"}))
+                         .union(Set({address: a, state: "lost"}))
+        s.with("connections", y)
+    }
+
+ // return DialPeerSuccess -> if connection.address.id not in Peers -> peerState(address) = "outbound" 
+ //                           otherwise -> delete connection from map 
+ // in the spec right now: the peer might not exists, be already outbound (stutter), be reconnecting
+ // TODO: check the cases
+    action lowerLevelDialPeerSuccess(node: Address, a: Address): bool = all {
+        require(config.get(node).connections.contains({address: a, state: "outgoing"})), 
+        if (config.get(node).Peers().filter(x => x.ID == a.ID) == Set()) all {
+            // If I don't have the ID in my peers already, add the address
+            config' = config.set(node, config.get(node).putPeer(a, "outbound"))
+        }
+        else {
+            // I have the ID already. That is, it might be incoming, or I might have dialed over a different
+            // (network) address
+            config' = config.set(node, config.get(node).removeConnectionOutgoing(a))
```

Comment:
This is not a successful connection.

Reply:
Although this appears to be needed in order to remove the peer from the dialing state.